include/nn/bcat.h

```diff
@@ -0,0 +1,109 @@
+/**
+ * @file bcat.h
+ * @brief BCAT service implementation.
+ */
+
+#pragma once
+
+#include "types.h"
+#include "os.h"
+
+namespace nn {
+
+    namespace bcat {
+
+        struct DirectoryName {
+            void isValid();
+        };
+
+        struct FileName {
+            void isValid(); 
+        };
+
+        class DeliveryCacheDirectory {
+            DeliveryCacheDirectory();
+            ~DeliveryCacheDirectory();
+            //void Open(char const*);
+            Result Open(nn::bcat::DirectoryName const&);
+            void GetCount();
+            void Close();
+        }; // namespace DeliveryCacheDirectory
+
+        class DeliveryCacheFile {
+            DeliveryCacheFile();
+            ~DeliveryCacheFile();
+            Result Open(char const*);
+            void Read(u64* file1, s64, void *, u64 file2); // probably better names
+            Result GetSize();
+            Result GetDigest();
+            void Close();
+        }; // namespace DeliveryCacheFile
+
+        class DeliveryCacheProgress {
+            DeliveryCacheProgress();
+            ~DeliveryCacheProgress();
+            void Detach();
+            void Update();
+            Result GetStatus();
+            Result GetCurrentDirectoryName();
+            Result GetCurrentFileName();
+            Result GetCurrentDownloaded();
+            Result GetCurrentTotale();
+            Result GetWholeDownloaded();
+            Result GetWholeTotal();
+            void GetResult();
+            // void Attach(nn::bcat::detail::ipc::IDeliveryCacheProgressService*);
+        }; // namespace DeliveryCacheProgress
+
+        namespace detail {
```

Comment:
I'm not sure it's worth adding `detail` unless it's observable from the public interface.